cmd/access/cluster.go

```diff
@@ -0,0 +1,446 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	clientcmdapiv1 "k8s.io/client-go/tools/clientcmd/api/v1"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// impersonateUser represents the user SREs are allowed to impersonate in order to retrieve a cluster's kubeconfig.
+	impersonateUser     = "backplane-cluster-admin"
+	kubeconfigSecretKey = "kubeconfig"
+
+	// PrivateLink "jump pod" configuration
+	jumpImage         = "image-registry.openshift-image-registry.svc:5000/openshift/cli:latest"
+	jumpContainerName = "jump"
+	jumpPodLabelKey   = "automated-break-glass-access/cluster"
+
+	// Lifespan for jump pods in seconds. Currently, PrivateLink jump pods will expire after 8 hours
+	jumpPodLifespan = 28800
+)
+
+var (
+	jumpPodPollInterval = 5 * time.Second
+	jumpPodPollTimeout  = 5 * time.Minute
+)
+
+// newCmdCluster implements the 'cluster' access subcommand
+func newCmdCluster(streams genericclioptions.IOStreams, flags *genericclioptions.ConfigFlags) *cobra.Command {
+	clusterCmd := &cobra.Command{
+		Use:               "cluster <cluster identifier>",
+		Short:             "Emergency access to a cluster",
+		Long:              "Obtain emergency credentials to access the given cluster. You must be logged into the cluster's hive shard",
+		Args:              cobra.ExactArgs(1),
+		DisableAutoGenTag: true,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(clusterCmdComplete(cmd, args))
+			// Prior to creating k8s client, verify the user has elevated permissions
+			cmdutil.CheckErr(verifyPermissions(streams, flags))
+			client := k8s.NewClient(flags)
+			clusterAccess := newClusterAccessOptions(client, streams, flags)
+			cmdutil.CheckErr(clusterAccess.Run(cmd, args))
+		},
+	}
+	return clusterCmd
+}
+
+// clusterCmdComplete verifies the command's invocation, returning an error if the usage is invalid
+func clusterCmdComplete(cmd *cobra.Command, args []string) error {
+	if len(args) != 1 {
+		return cmdutil.UsageErrorf(cmd, "Exactly one cluster identifier was expected")
+	}
+	return osdctlutil.IsValidClusterKey(args[0])
+}
+
+// verifyPermissions determines if the user has supplied the correct permissions in order to retrieve a KubeConfig secret from hive.
+// If the user attempts to impersonate an unexpected account, an error is returned.
+// If the user hasn't attempted to impersonate anyone, it prompts whether they would like to do so automatically.
+func verifyPermissions(streams genericclioptions.IOStreams, flags *genericclioptions.ConfigFlags) error {
+	if flags.Impersonate != nil && *flags.Impersonate != "" {
+		if *flags.Impersonate != impersonateUser {
+			osdctlutil.StreamPrintln(streams, "")
+			return fmt.Errorf("Unauthorized impersonation as user '%s'. Only requests to impersonate '%s' are allowed.", *flags.Impersonate, impersonateUser)
+		}
+	} else {
```

Comment:
Else is required in the case where the `flags.Impersonate` field is already set to the correct `impersonateUser` (`backplane-cluster-admin`); removing it would run the prompt logic when the correct user has already been set.